File: python/pii/span_mapper.py
from __future__ import annotations

from dataclasses import dataclass

from pii.classifier import PiiSpan
from pii.ocr import OcrToken

@dataclass
class LabeledRect:
    left: int
    top: int
    right: int
    bottom: int
    label: str

def _overlaps(token: OcrToken, span: PiiSpan) -> bool:
    return token.start < span.end_exclusive and token.end_exclusive > span.start

def _vertical_overlap_ratio(r1_top: int, r1_bottom: int, r2_top: int, r2_bottom: int) -> float:
    overlap_top = max(r1_top, r2_top)
    overlap_bottom = min(r1_bottom, r2_bottom)
    if overlap_bottom <= overlap_top:
        return 0.0
    overlap_height = overlap_bottom - overlap_top
    min_height = min(r1_bottom - r1_top, r2_bottom - r2_top)
    if min_height <= 0:
        return 0.0
    return overlap_height / min_height

def map_spans(
    tokens: list[OcrToken],
    spans: list[PiiSpan],
    image_width: int,
    image_height: int,
) -> list[LabeledRect]:
    rects: list[LabeledRect] = []

    for span in spans:
        matching_tokens = [t for t in tokens if _overlaps(t, span)]
        if not matching_tokens:
            continue

        token_rects = [
            (t.left, t.top, t.right, t.bottom)
            for t in matching_tokens
        ]
        token_rects.sort(key=lambda r: (r[1], r[0]))

        merged: list[list[int]] = []
        for left, top, right, bottom in token_rects:
            if merged:
                prev = merged[-1]
                h_gap = left - prev[2]
                v_ratio = _vertical_overlap_ratio(prev[1], prev[3], top, bottom)
                if h_gap < 8 and v_ratio > 0.5:
                    prev[0] = min(prev[0], left)
                    prev[1] = min(prev[1], top)
                    prev[2] = max(prev[2], right)
                    prev[3] = max(prev[3], bottom)
                    continue
            merged.append([left, top, right, bottom])

        for m in merged:
            pad = 4
            r_left = max(0, m[0] - pad)
            r_top = max(0, m[1] - pad)
            r_right = min(image_width, m[2] + pad)
            r_bottom = min(image_height, m[3] + pad)
            rects.append(LabeledRect(
                left=r_left,
                top=r_top,
                right=r_right,
                bottom=r_bottom,
                label=span.label,
            ))

    return rects

File: rust/src/pii/span_mapper.rs
use serde::{Deserialize, Serialize};

use super::ocr::OcrToken;
use super::spans::PiiSpan;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-bindings", derive(ts_rs::TS), ts(export))]
pub struct LabeledRect {
    pub left: i32,
    pub top: i32,
    pub right: i32,
    pub bottom: i32,
    pub label: String,
}

pub fn map_spans(
    tokens: &[OcrToken],
    spans: &[PiiSpan],
    img_width: u32,
    img_height: u32,
) -> Vec<LabeledRect> {
    let mut result = Vec::new();

    for span in spans {
        let overlapping: Vec<&OcrToken> = tokens
            .iter()
            .filter(|t| t.start < span.end_exclusive && t.end_exclusive > span.start)
            .collect();

        if overlapping.is_empty() {
            continue;
        }

        let mut rects: Vec<(i32, i32, i32, i32)> = overlapping
            .iter()
            .map(|t| (t.rect.left, t.rect.top, t.rect.right, t.rect.bottom))
            .collect();

        rects.sort_by(|a, b| a.1.cmp(&b.1).then(a.0.cmp(&b.0)));

        let merged = merge_rects(&rects);

        for (left, top, right, bottom) in merged {
            let left = (left - 4).max(0);
            let top = (top - 4).max(0);
            let right = (right + 4).min(img_width as i32);
            let bottom = (bottom + 4).min(img_height as i32);

            result.push(LabeledRect {
                left,
                top,
                right,
                bottom,
                label: span.label.clone(),
            });
        }
    }

    result
}

fn merge_rects(rects: &[(i32, i32, i32, i32)]) -> Vec<(i32, i32, i32, i32)> {
    if rects.is_empty() {
        return Vec::new();
    }

    let mut merged: Vec<(i32, i32, i32, i32)> = Vec::new();
    merged.push(rects[0]);

    for &(left, top, right, bottom) in &rects[1..] {
        let last = merged.last_mut().unwrap();
        let h_gap = left - last.2;
        let min_height = (last.3 - last.1).min(bottom - top);
        let overlap_top = last.1.max(top);
        let overlap_bottom = last.3.min(bottom);
        let v_overlap = (overlap_bottom - overlap_top).max(0);

        let should_merge = h_gap < 8 && min_height > 0 && v_overlap * 2 > min_height;

        if should_merge {
            last.0 = last.0.min(left);
            last.1 = last.1.min(top);
            last.2 = last.2.max(right);
            last.3 = last.3.max(bottom);
        } else {
            merged.push((left, top, right, bottom));
        }
    }

    merged
}
